File: src/graph/algorithms.rs
// src_new/graph/algorithms.rs

// Algorithm implementations directly on FastGraph for MVP
use crate::graph::core::FastGraph;
use crate::graph::types::{NodeId, EdgeId};
use std::collections::{VecDeque, HashSet, HashMap};

impl FastGraph {
    /// Breadth-First Search (BFS) traversal from a given node.
    /// Returns the order of visited node IDs.
    pub fn bfs(&self, start: NodeId) -> Vec<NodeId> {
        let mut visited = HashSet::new();
        let mut queue = VecDeque::new();
        let mut order = Vec::new();
        queue.push_back(start.clone());
        visited.insert(start.clone());
        while let Some(node) = queue.pop_front() {
            order.push(node.clone());
            for edge in &self.edge_collection.edge_ids {
                if edge.0 == node && !visited.contains(&edge.1) {
                    visited.insert(edge.1.clone());
                    queue.push_back(edge.1.clone());
                }
            }
        }
        order
    }

    /// Depth-First Search (DFS) traversal from a given node.
    /// Returns the order of visited node IDs.
    pub fn dfs(&self, start: NodeId) -> Vec<NodeId> {
        let mut visited = HashSet::new();
        let mut stack = Vec::new();
        let mut order = Vec::new();
        stack.push(start.clone());
        while let Some(node) = stack.pop() {
            if !visited.insert(node.clone()) {
                continue;
            }
            order.push(node.clone());
            for edge in &self.edge_collection.edge_ids {
                if edge.0 == node && !visited.contains(&edge.1) {
                    stack.push(edge.1.clone());
                }
            }
        }
        order
    }

    /// Finds the shortest path (unweighted) between two nodes using BFS.
    /// Returns the path as a vector of NodeId, or empty if unreachable.
    pub fn shortest_path(&self, start: NodeId, goal: NodeId) -> Vec<NodeId> {
        let mut visited = HashSet::new();
        let mut queue = VecDeque::new();
        let mut parent: HashMap<NodeId, NodeId> = HashMap::new();
        queue.push_back(start.clone());
        visited.insert(start.clone());
        while let Some(node) = queue.pop_front() {
            if node == goal {
                // Reconstruct path
                let mut path = vec![goal.clone()];
                let mut curr = &goal;
                while let Some(p) = parent.get(curr) {
                    path.push(p.clone());
                    curr = p;
                }
                path.reverse();
                return path;
            }
            for edge in &self.edge_collection.edge_ids {
                if edge.0 == node && !visited.contains(&edge.1) {
                    visited.insert(edge.1.clone());
                    parent.insert(edge.1.clone(), node.clone());
                    queue.push_back(edge.1.clone());
                }
            }
        }
        Vec::new()
    }

    /// Finds all connected components (for undirected graphs).
    /// Returns a vector of components, each a vector of NodeId.
    pub fn connected_components(&self) -> Vec<Vec<NodeId>> {
        let mut visited = HashSet::new();
        let mut components = Vec::new();
        for node in &self.node_collection.node_ids {
            if !visited.contains(node) {
                let mut stack = vec![node.clone()];
                let mut component = Vec::new();
                while let Some(n) = stack.pop() {
                    if visited.insert(n.clone()) {
                        component.push(n.clone());
                        for edge in &self.edge_collection.edge_ids {
                            if edge.0 == n && !visited.contains(&edge.1) {
                                stack.push(edge.1.clone());
                            }
                            if edge.1 == n && !visited.contains(&edge.0) {
                                stack.push(edge.0.clone());
                            }
                        }
                    }
                }
                components.push(component);
            }
        }
        components
    }

    /// Computes the clustering coefficient for each node (simple version).
    /// Returns a map from NodeId to clustering coefficient (f64).
    pub fn clustering_coefficient(&self) -> HashMap<NodeId, f64> {
        let mut coeffs = HashMap::new();
        let nodes = &self.node_collection.node_ids;
        let edges = &self.edge_collection.edge_ids;
        let mut neighbors: HashMap<&NodeId, HashSet<&NodeId>> = HashMap::new();
        for node in nodes {
            neighbors.insert(node, HashSet::new());
        }
        for edge in edges {
            neighbors.get_mut(&edge.0).map(|n| { n.insert(&edge.1); });
            neighbors.get_mut(&edge.1).map(|n| { n.insert(&edge.0); });
        }
        for node in nodes {
            let neigh: Vec<&NodeId> = neighbors.get(node).unwrap().iter().cloned().collect();
            let k = neigh.len();
            if k < 2 {
                coeffs.insert(node.clone(), 0.0);
                continue;
            }
            let mut links = 0;
            for i in 0..k {
                for j in (i+1)..k {
                    if neighbors.get(neigh[i]).unwrap().contains(neigh[j]) {
                        links += 1;
                    }
                }
            }
            let denom = (k * (k - 1)) / 2;
            coeffs.insert(node.clone(), links as f64 / denom as f64);
        }
        coeffs
    }
}

File: python/groggy/algorithms.py
# python_new/groggy/algorithms.py

def get_nodes_attribute(graph, node_ids, attr_name):
    """
    Get a specific attribute for multiple nodes efficiently (Rust backend).
    Args:
        graph (Graph): The graph to query.
        node_ids (list): List of node IDs.
        attr_name (str): Name of the attribute to retrieve.
    Returns:
        dict: Mapping node ID to attribute value.
    """
    return graph.nodes.attr().get(node_ids, attr_name)

def get_nodes_attributes(graph, node_ids):
    """
    Get all attributes for multiple nodes efficiently (Rust backend).
    Args:
        graph (Graph): The graph to query.
        node_ids (list): List of node IDs.
    Returns:
        dict: Mapping node ID to attribute dict.
    """
    return graph.nodes.attr().get(node_ids)

def get_edges_attribute(graph, edge_ids, attr_name):
    """
    Get a specific attribute for multiple edges efficiently (Rust backend).
    Args:
        graph (Graph): The graph to query.
        edge_ids (list): List of edge IDs.
        attr_name (str): Name of the attribute to retrieve.
    Returns:
        dict: Mapping edge ID to attribute value.
    """
    return graph.edges.attr().get(edge_ids, attr_name)

def get_edges_attributes(graph, edge_ids):
    """
    Get all attributes for multiple edges efficiently (Rust backend).
    Args:
        graph (Graph): The graph to query.
        edge_ids (list): List of edge IDs.
    Returns:
        dict: Mapping edge ID to attribute dict.
    """
    return graph.edges.attr().get(edge_ids)

def bfs(graph, start_node, **kwargs):
    """
    Performs a breadth-first search (BFS) traversal starting from a given node.
    
    Delegates traversal logic to backend or optimized core. Supports early stopping, filtering, and batch traversal.
    Args:
        graph (Graph): The graph to traverse.
        start_node: Node ID to start from.
        **kwargs: Optional parameters (e.g., max_depth, filter).
    Returns:
        list: List of visited node IDs in BFS order.
    Raises:
        KeyError: If start_node not found.
    """
    # TODO: 1. Validate input; 2. Delegate to backend/core; 3. Support options.
    pass

def dfs(graph, start_node, **kwargs):
    """
    Performs a depth-first search (DFS) traversal starting from a given node.
    
    Delegates traversal logic to backend or optimized core. Supports early stopping, filtering, and batch traversal.
    Args:
        graph (Graph): The graph to traverse.
        start_node: Node ID to start from.
        **kwargs: Optional parameters (e.g., max_depth, filter).
    Returns:
        list: List of visited node IDs in DFS order.
    Raises:
        KeyError: If start_node not found.
    """
    # TODO: 1. Validate input; 2. Delegate to backend/core; 3. Support options.
    pass

def shortest_path(graph, source, target, **kwargs):
    """
    Computes the shortest path between two nodes using the selected algorithm.
    
    Delegates to backend or core for Dijkstra, BFS, or other algorithms. Supports weighted/unweighted graphs and custom filters.
    Args:
        graph (Graph): The graph to search.
        source: Source node ID.
        target: Target node ID.
        **kwargs: Optional parameters (e.g., method, weight_attr).
    Returns:
        list: Sequence of node IDs forming the shortest path, or None if unreachable.
    Raises:
        KeyError: If source or target not found.
    """
    if not graph.nodes.has(source):
        raise KeyError(f"Source node {source} not found.")
    if not graph.nodes.has(target):
        raise KeyError(f"Target node {target} not found.")
    return graph._rust.shortest_path(source, target, **kwargs)

def connected_components(graph):
    """
    Finds all connected components in the graph.
    
    Delegates to backend for efficient batch traversal. Supports both directed and undirected graphs.
    Args:
        graph (Graph): The graph to analyze.
    Returns:
        list: List of components, each as a set of node IDs.
    """
    return graph._rust.connected_components()

def clustering_coefficient(graph, node=None):
    """
    Calculates the clustering coefficient for the graph or a specific node.
    
    Delegates to backend for vectorized computation. Supports batch mode for all nodes.
    Args:
        graph (Graph): The graph to analyze.
        node (optional): Node ID to compute for, or None for all nodes.
    Returns:
        float or dict: Coefficient value(s).
    Raises:
        KeyError: If node not found.
    """
    if node is not None and not graph.nodes.has(node):
        raise KeyError(f"Node {node} not found.")
    return graph._rust.clustering_coefficient(node) if node is not None else graph._rust.clustering_coefficient_all()

def pagerank(graph, **kwargs):
    """
    Computes PageRank scores for all nodes in the graph.
    
    Delegates to backend for efficient iterative computation. Supports weighted edges and convergence criteria.
    Args:
        graph (Graph): The graph to analyze.
        **kwargs: Optional parameters (e.g., damping, max_iter, tol).
    Returns:
        dict: Node ID to PageRank score mapping.
    """
    return graph._rust.pagerank(**kwargs)

def betweenness_centrality(graph, **kwargs):
    """
    Computes betweenness centrality for all nodes or edges in the graph.
    
    Delegates to backend for efficient batch computation. Supports normalization and directed/undirected graphs.
    Args:
        graph (Graph): The graph to analyze.
        **kwargs: Optional parameters (e.g., normalized, endpoints).
    Returns:
        dict: Node or edge ID to centrality score mapping.
    """
    return graph._rust.betweenness_centrality(**kwargs)

def label_propagation(graph, **kwargs):
    """
    Performs community detection using the label propagation algorithm.
    
    Delegates to backend for iterative, batch label updates. Supports custom initialization and stopping criteria.
    Args:
        graph (Graph): The graph to analyze.
        **kwargs: Optional parameters (e.g., max_iter, seed).
    Returns:
        dict: Node ID to label mapping.
    """
    return graph._rust.label_propagation(**kwargs)

def louvain(graph, **kwargs):
    """
    Detects communities using the Louvain modularity maximization algorithm.
    
    Delegates to backend for efficient modularity optimization. Supports weighted graphs and multilevel refinement.
    Args:
        graph (Graph): The graph to analyze.
        **kwargs: Optional parameters (e.g., weight_attr, max_iter).
    Returns:
        dict: Node ID to community label mapping.
    """
    return graph._rust.louvain(**kwargs)

def modularity(graph, **kwargs):
    """
    Calculates the modularity score of a given labeling or partitioning.
    
    Delegates to backend for efficient modularity computation. Supports weighted and directed graphs.
    Args:
        graph (Graph): The graph to analyze.
        **kwargs: Optional parameters (e.g., labels, weight_attr).
    Returns:
        float: Modularity score.
    """
    return graph._rust.modularity(**kwargs)
